test/exe/Main.hs

```diff
@@ -585,6 +586,123 @@ addSigActionTests = let
   , "a `haha` b = a b"       >:: "haha :: (t1 -> t2) -> t1 -> t2"
   ]
 
+findDefinitionTests :: TestTree
+findDefinitionTests = let
+
+  tst (get, check) pos targetRange title = testSession title $ do
+    doc <- openDoc' "Testing.hs" "haskell" source
+    found <- get doc pos
+    check found targetRange
+
+  checkDefs defs expected = do
+
+    let ndef = length defs
+    if ndef /= 1
+      then let dfound n = "definitions found: " <> show n in
+           liftIO $ dfound (1 :: Int) @=? dfound (length defs)
+      else do
+           let [Location{_range = foundRange}] = defs
+           liftIO $ expected @=? foundRange
+
+  checkHover hover expected = do
+    case hover of
+      Nothing -> liftIO $ "hover found" @=? ("no hover found" :: T.Text)
+      Just Hover{_contents = (HoverContents MarkupContent{_value = v})} ->
+        liftIO $ adjust expected @=? Position l c where
+          found = T.splitOn ":" $ head $ T.splitOn "**" $ last $ T.splitOn "Testing.hs:" v
+          [l,c] =  map (read . T.unpack) found
+          -- looks like hovers use 1-based numbering while definitions use 0-based
```

Comment:
good catch, that sounds like a bug.

Reply:
Which do we prefer, 0 or 1 based?

Reply:
LSP expects both lines and columns to be 0-based so we should follow that.

Reply:
These 1-based numbers come from GHC itself, specifically [`SrcLoc`](https://hackage.haskell.org/package/ghc-8.6.5/docs/src/SrcLoc.html#RealSrcLoc).

They only appear in the hover message text, rather than being used programatically anywhere (except for these tests), so I wonder to what extent it is important.